scripts/ASTImportTest.sh

```diff
@@ -1,105 +1,192 @@
+    cat "$stdout_file" >&2
+}
+
+IMPORT_TEST_TYPE=
+EXIT_ON_ERROR=0
+for PARAM in "$@"
+do
+    case "$PARAM" in
+        ast) IMPORT_TEST_TYPE="ast" ;;
+        --exit-on-error) EXIT_ON_ERROR=1 ;;
+        *) print_usage ;;
+    esac
+done
 
 SYNTAXTESTS_DIR="${REPO_ROOT}/test/libsolidity/syntaxTests"
 ASTJSONTESTS_DIR="${REPO_ROOT}/test/libsolidity/ASTJSON"
-NSOURCES="$(find "$SYNTAXTESTS_DIR" -type f | wc -l)"
-
-# DEV_DIR="${REPO_ROOT}/../tmp/contracts/"
-# NSOURCES="$(find $DEV_DIR -type f | wc -l)" #TODO use find command
 
 FAILED=0
 UNCOMPILABLE=0
 TESTED=0
 
-if [[ "$(find . -maxdepth 0 -type d -empty)" == "" ]]; then
-    echo "Test directory not empty. Skipping!"
-    exit 1
-fi
+function ast_import_export_equivalence
+{
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
 
-# function tests whether exporting and importing again leaves the JSON ast unchanged
-# Results are recorded by adding to FAILED or UNCOMPILABLE.
-# Also, in case of a mismatch a diff and the respective ASTs are printed
-# Expected parameters:
-# $1 name of the file to be exported and imported
-# $2 any files needed to do so that might be in parent directories
-function testImportExportEquivalence {
-    local nth_input_file="$1"
-    IFS=" " read -r -a all_input_files <<< "$2"
+    local export_command=("$SOLC" --combined-json ast --pretty-json --json-indent 4 "${input_files[@]}")
+    local import_command=("$SOLC" --import-ast --combined-json ast --pretty-json --json-indent 4 expected.json)
 
-    if $SOLC "$nth_input_file" "${all_input_files[@]}" > /dev/null 2>&1
+    # export ast - save ast json as expected result (silently)
+    if ! "${export_command[@]}" > expected.json 2> stderr_export.txt
     then
-        ! [[ -e stderr.txt ]] || { echo "stderr.txt already exists. Refusing to overwrite."; exit 1; }
+        print_stderr_stdout "ERROR: AST reimport failed (export) for input file ${sol_file}." ./stderr_export.txt ./expected.json
+        print_used_commands "$(pwd)" "${export_command[*]} > expected.json" "${import_command[*]}"
+        return 1
+    fi
 
-        # save exported json as expected result (silently)
-        $SOLC --combined-json ast --pretty-json "$nth_input_file" "${all_input_files[@]}" > expected.json 2> /dev/null
-        # import it, and export it again as obtained result (silently)
-        if ! $SOLC --import-ast --combined-json ast --pretty-json expected.json > obtained.json 2> stderr.txt
+    # (re)import ast - and export it again as obtained result (silently)
+    if ! "${import_command[@]}" > obtained.json 2> stderr_import.txt
+    then
+        print_stderr_stdout "ERROR: AST reimport failed (import) for input file ${sol_file}." ./stderr_import.txt ./obtained.json
+        print_used_commands "$(pwd)" "${export_command[*]} > expected.json" "${import_command[*]}"
+        return 1
+    fi
+
+    # compare expected and obtained ast's
+    if diff_files expected.json obtained.json
+    then
+        echo -n "✅"
```

Comment:
After staring on these nice unicode checkmarks for some hours today, I think they should not be used ;)

Reply:
Maybe we could ignore the print in case of success and only print an error if they are not equivalent?

Reply:
Yes, the normal dots are fine in my opinion. I was thinking of removing only the checkmarks as you mentioned. And keep the error in case the ast's are not equivalent.